File: labs/lab3/EchoServer.go
/* Simple EchoServer in GoLang by Phu Phung, customized by Riley Miranda for SecAD*/
package main

import (
	"fmt"
	"net"
	"os"
)

const BUFFERSIZE int = 1024
var allClient_conns = make(map[net.Conn]string)
func main() {
	if len(os.Args) != 2 {
		fmt.Printf("Usage: %s <port>\n", os.Args[0])
		os.Exit(0)
	}
	port := os.Args[1]
	if len(port) > 5 {
		fmt.Println("Invalid port value. Try again!")
		os.Exit(1)
	}
	server, err := net.Listen("tcp", ":"+port)
	if err != nil {
		fmt.Printf("Cannot listen on port '" + port + "'!\n")
		os.Exit(2)
	}
	fmt.Println("EchoServer in GoLang developed by Phu Phung, SecAD, revised by Your Name")
	fmt.Printf("EchoServer is listening on port '%s' ...\n", port)
	for {
		client_conn, _ := server.Accept()
		go client_goroutine(client_conn)
	}
}

func client_goroutine(client_conn net.Conn){
	fmt.Printf("A new client '%s' connected!\n", client_conn.RemoteAddr().String())
	allClient_conns[client_conn] = client_conn.RemoteAddr().String()
	fmt.Printf("# of connected clients: %d\n", len(allClient_conns))
	var buffer [BUFFERSIZE]byte
	for {
		byte_received, read_err := client_conn.Read(buffer[0:])
		if read_err != nil {
			fmt.Println("Error in receiving...")
			return
		}
		_, write_err := client_conn.Write(buffer[0:byte_received])
		if write_err != nil {
			fmt.Println("Error in sending...")
			return
		}
		fmt.Printf("Received data: %sEchoed back!\n", buffer)
	}
}

File: labs/lab3/EchoServer.java
import java.net.*;
import java.io.*;
import java.util.ArrayList;

public class EchoServer {
    static ThreadList threadlist = new ThreadList();

    public static void main(String[] args) throws IOException {
        if (args.length != 1) {
            System.err.println("Usage: java EchoServer <port number>");
            System.exit(1);
        }
        
        int portNumber = Integer.parseInt(args[0]);
        System.out.println("EchoServer by Riley Miranda");
        try {
            ServerSocket serverSocket =
                new ServerSocket(Integer.parseInt(args[0]));
            System.out.println("EchoServer is running at port " + Integer.parseInt(args[0]));
            while(true) {
                Socket clientSocket = serverSocket.accept(); 
                System.out.println("A client is connected ");    
                new EchoServerThread(threadlist, clientSocket).start();
            }
        } catch (IOException e) {
            System.out.println("Exception caught when trying to listen on port "
                + portNumber + " or listening for a connection");
            System.out.println(e.getMessage());
        }
    }
}
class EchoServerThread extends Thread{
    private Socket clientSocket = null;
    private ThreadList threadlist = null;
    public EchoServerThread(ThreadList threadlist, Socket socket) {
        clientSocket = socket;
        this.threadlist = threadlist;
    }
    public void run() {
        System.out.println("A new thread for client is running...");
        if(threadlist!=null) {
            threadlist.addThread(this);
            System.out.println("Inside thread: total clients: " + 
                threadlist.getNumberofThreads());
        }
        try {
            PrintWriter out =
                    new PrintWriter(clientSocket.getOutputStream(), true);                   
            BufferedReader in = new BufferedReader(
                    new InputStreamReader(clientSocket.getInputStream()));
        
            String inputLine;
            while ((inputLine = in.readLine()) != null) {
                System.out.println("received from client: " + inputLine);
                System.out.println("Echo back");
                out.println(inputLine);
            }
        } catch(IOException e) {
            System.out.println("IO Exception in thread" + e.getMessage());
        }
    }
}
class ThreadList{
    //private ... threadlist = //store the list of threads in this variable 
    private ArrayList<EchoServerThread> threadlist = new ArrayList<EchoServerThread>();
    public ThreadList(){        
    }
    public int getNumberofThreads(){
    //return the number of current threads
        return threadlist.size();
    }
    public void addThread(EchoServerThread newThread){
    //add the newthread object to the threadlist
    threadlist.add(newThread);    
    }
    public void removeThread(EchoServerThread thread){
    //remove the given thread from the threadlist 
    threadlist.remove(thread);      
    }
    public void sendToAll(String message){
    //ask each thread in the threadlist to send the given message to its client     
    }
}